README.md
The SleepCycles package
================

<!-- README.md is generated from README.Rmd. Please edit that file -->

## Package and main function

The SleepCycles package and its main function `SleepCycles()` is
designed to detect sleep cycles and their respective NREM and REM parts
(called (N)REM periods) from data that has been sleep staged according
to AASM criteria. Additionally, each (N)REM part is split into
percentiles. The function results in a plot visualising the results and
creates a text file so the results can be used for further processing.

### Detection of sleep cycles

Sleep cycles are largely detected according to the originally proposed
criteria by Feinberg & Floyd (1979) and as described in Blume & Cajochen
(2021) . NREM periods are periods starting with N1 (default) or N2 at
the beginning of the night and W or another NREM stage following a REM
period. NREMPs have a minimal duration minimal duration of 15min (can
include W, up to &lt;5min REM, except for the first REMP, for which
there is no minimum duration criterion). REM following a NREM period
always represents a potential REM period, however any REMP must be at
least 5min long (except the first REMP, for which no minimum duration
criterion is applied). If a NREMP exceeds 120min in duration (excl.
wake), it can be split into 2 parts. The new cycle starts with the first
N3 episode following a phase (&gt;12min) with any other stage than N3,
that is a lightening of sleep (cf. Rudzik et al., 2020; Jenni et al.,
2004; Kurth et al., 2010). The function makes suggestions where
splitting could be done according to these criteria and visualises the
potential splitting points on top of a hypnogram. The user can then
interactively choose where to split the NREMP. However, the code also
offers the possibility to provide a numeric value for an epoch at which
to split or you can also decide to not split at all. A combination of a
NREMP and the following REMP represents one sleep cycle, except for the
case when a NREMP is split. In this case, the first of the two resulting
NREMPs represents a sleep cycle (without REM).

### Requirements

The function requires any sleep staging results file with a column, in
which the sleep stages are coded in the usual numeric 0,1,2,3,5 (i.e.,
W, N1, N2, N3, REM) pattern (i.e., a numeric vector). The user can
define other integers to be handled as W or N3 (i.e. in the case
stagings were done according to the Rechtschaffen & Kales criteria
including S3 and S4). The presence of further columns, e.g. a ‘time’
column, is not an issue. Staging must be in 30s epochs. Besides text
files, the `SleepCycles()` function can also handle csv and marker files
for the Brain Vision Analyzer. The input file type can be indicated with
the `filetype` argument (`filetype = "txt"` (default) or
`filetype = "csv"` or `filetype = "vmrk"`).

#### Details

Besides sleep cycles (NREM-REM), the result also splits the NREM and REM
parts of each cycle in percentiles. In case the length of a period is
not divisible by 10 (e.g., 203 epochs), one epoch is added to
percentiles in a randomized fashion to reach the correct length of a
period (here: 7 percentiles of 20 epochs, 3 of 21 epochs).

The code offers to choose whether incomplete periods should be removed
at the end of the night (argument `rm_incomplete_period`, default =
FALSE). Incomplete periods are defined by periods that are followed by
&lt;5min NREM or W (e.g. because a participant is woken up).

Although this is not encouraged, for some participants it may be
necessary to decrease the minimum duration of REM from 5min to 4 or
4.5min as otherwise a seemingly ‘clear’ REM period is skipped. While the
default length of REMPs is 10 segments, it can be decreased.

The user can either process all files in a given directory (default) or
specific files by specifying a vector of files (argument `files`).

By default, the function produces and saves a plot for visual inspection
of the results (argument `plot`, default = TRUE).

#### Arguments in SleepCycle function

`p` character vector indicating the directory containing the sleep
staging files  
`sleepstart` character vector indicating whether the first NREMP at the
beginning of the night should start with N1 or N2. Default: N1 `files`
numeric vector indicating which files in ‘p’ to process. Default: NA  
`filetype` character indicating file type of the files containing the
sleep staging results. Can be “txt” (default) or “csv” or “vmrk” (i.e.,
marker files for Brain Vision Analyzer Software).  
`treat_as_W` numeric vector indicating which values should be treated as
‘wake’. Default: NA  
`treat_as_N3` numeric vector indicating which values should be treated
as ‘N3’. Default: NA  
`rm_incomplete_period` logical: should incomplete periods at the end of
the night be removed? Default: F.  
`plot` logical: should a plot for the result of the detection procedure
be generated and saved? Default: T.  
`REMP_length` numeric value specifying the minimum duration of a REM
period. Default is 10 segments (i.e. 5 minutes). Decreasing the min.
length is not encouraged and should only be done following careful
consideration

### Worked example

First, we install and load the package if we haven’t done so.

``` r
## First, we save your current workspace
save.image(file=paste(tempdir(), "currsession.RData", sep = "/"))
## make sure you start with a clean session.
rm(list = ls(all = TRUE))
install.packages("SleepCycles", repos = "http://cran.us.r-project.org")
```

Then, we are ready to use the package on our data set. Note that in the
`sleepstages2` data set, the first NREM period exceeds 120 minutes.
Thus, the code attempts to split this NREM period.

The text file has a header, thus, when asked whether it has a header
file, type `y`. Columns are separated by comma, thus type `,` when
prompted. Of course, if we had several of these files in our directory,
they would all have to have the same pattern.

##### Create directory & save data file

First, we load the `sleepstages2` file that comes with the package and
create a directory, where we save it.

``` r
library(SleepCycles)
data(sleepstages2)

## save current working directory so we can reset this later.
olddir <- getwd()

## create a new directory in the temporary directory (don't worry, it will automatically be deleted  
## when you restart your computer)
newdir <- file.path(tempdir(),"SleepCycles_exmpl2")
dir.create(newdir, showWarnings = FALSE)

## write the sleepstages2 file to this new directory
write.table(sleepstages2, file = paste(newdir, "sleepstages2.txt", sep = "/"),
row.names=FALSE, col.names = TRUE, quote = FALSE, sep = ",")
```

##### Run the detection

Then, we apply the actual `SleepCycles` function. The file contains
column names in a header and columns are separated with a comma. When we
are prompted, we have to decide where we want to split the data, either
at the first or the second suggested location. I would suggest selecting
the fist, so when prompted, we simply type `1`.

``` r
SleepCycles::SleepCycles(newdir, filetype = "txt")

## We again load the workspace image from before the code above was executed
save.image(file=paste(tempdir(), "currsession.RData", sep = "/"))

## we set the directory back to the one we were using before as we were just working in the  
## temp directory.
setwd(olddir)
```


NEWS.md
The SleepCycles package
================

<!-- NEWS.md is generated from NEWS.Rmd. Please edit that file -->

## News

### SleepCycles 1.1.4 (2021-22-09)

Solved an issue that could result in problems with vignette building.

### SleepCycles 1.1.3 (2021-11-08)

Code in the Readme file that violated CRAN policy has been replaced.

### SleepCycles 1.1.2 (2021-25-04)

Enhancement: now also files in which a sleep stage (e.g. N3) is missing
can be read in.

### SleepCycles 1.1.1 (2021-29-03)

Updated the reference and doi to the companion paper by Blume &
Cajochen. Created a CITATION file for the publication.

### SleepCycles 1.1.0 (2021-18-02)

Bug fix: the load\_prep.R function resulted in an error when input files
had less than two columns, this has been fixed. Bug fix: in the unlikely
event that there is only one REM sleep episode, the code resulted in an
error. This has been fixed. Enhancement: the user can now also specify
csv files as input files to the SleepCycles() function. Enhancement: the
user can now specify in the ‘sleepstart’ argument whether the first
NREMP of a night should start with N1 (default) or N2.

### SleepCycles 1.0.1 (2020-02-12)

No changes since version 0.8. This is the first version available on
CRAN.

### SleepCycles 1.0.0 (2020-24-11)

No changes since version 0.8. This is the first version submitted to
CRAN.

### SleepCycles 0.8 (2020-18-11)

First NEWS update. All changes will be documented in this file from now
on. This is also the first version available on CRAN.


R/data.R
#' Sleepstaging from one participant.
#'
#' A dataset containing sleep staging result for one participant. In this case, no NREM period exceeds 120min (excl. wake), wherefore no splitting is initiated.
#' The variables are as follows:
#'
#' @format A vmrk file (can be read as a text file) with 954 rows and 2 variables:
#' \describe{
#'   \item{Type}{Describes what column 'Description' shows}
#'   \item{Description}{Gives the sleep stage according to AASM criteria. 0 = wake, 1 = N1, 2 = N2, 3 = N3, 5 = W}
#' }
"sleepstages"

#' Sleepstaging from one participant.
#'
#' A dataset containing sleep staging result for one participant. In this case, a NREM period exceeds 120min (excl. wake), wherefore splitting is initiated.
#' The variables are as follows:
#'
#' @format A vmrk file (can be read as a text file) with 959 rows (including one header row) and 2 variables:
#' \describe{
#'   \item{Type}{Describes what column 'Description' shows}
#'   \item{Description}{Gives the sleep stage according to AASM criteria. 0 = wake, 1 = N1, 2 = N2, 3 = N3, 5 = W}
#' }
"sleepstages2"
NULL


R/split.R
split <- list(

#' @description Auxiliary function. Checks if any NREM part (excluding Wake) of a NREM period is > 120 minutes.
  is.toolong <- function(data){
    # check if the first period is actually a REM period
    NREM1 <- which(data$CycleStart == "NREMP")
    REM1 <- which(data$CycleStart == "REMP")
    
    if(REM1[1] < NREM1[1]){
      ## REMP precedes first NREMP (i.e., REM onset)
      message("ATTENTION: first REM period precedes NREM period!")
      
      ## is any NREM part (excl. wake) of a NREMP longer than 120min?
      cycs <- which(data$CycleStart == "NREMP" | data$CycleStart == "REMP")
      
      toolong <- NA
      for (k in seq(3,length(cycs),2)){ #check every second one as only every second is a NREMP
        subset <- data[c(cycs[k-1]:(cycs[k]-1)),]
        wake_eps <- sum(subset$Descr2 == "W")
        if (all(is.na(wake_eps))){
          wake_eps <- 0 #set 0 to avoid error in case none is found
        } 
        if (((cycs[k]-cycs[k-1])-wake_eps)>=240){ #>= as cycs[2] (cf. line 7) is already the beginning of a REMP
          toolong <- c(toolong, cycs[k-1])
        }
      }
      toolong <- na.omit(toolong)
    }else{
      ## NREMP precedes first REMP as expected
      ## is any NREM part (excl. wake) of a NREMP longer than 120min?
      cycs <- which(data$CycleStart == "NREMP" | data$CycleStart == "REMP")
      
      toolong <- NA
      for (k in seq(2,length(cycs),2)){ #check every second one as only every second is a NREMP
        subset <- data[c(cycs[k-1]:(cycs[k]-1)),]
        wake_eps <- sum(subset$Descr2 == "W")
        if (all(is.na(wake_eps))){
          wake_eps <- 0 #set 0 to avoid error in case none is found
        } 
        if (((cycs[k]-cycs[k-1])-wake_eps)>=240){ #>= as cycs[2] (cf. line 7) is already the beginning of a REMP
          toolong <- c(toolong, cycs[k-1])
        }
      }
      toolong <- na.omit(toolong)
    }
    return(toolong)
  },
  
#' @description   Auxiliary function. Splits NREMPs that are too long (i.e., > 120 min). The function
#' makes suggestions for splitting at N3 epoch(s) following "lightening" of sleep. The user can also
#' decide to not split, enter a pre-defined epoch number to split at, or skip the night.
  
toolong_split <- function(data, toolong, filename){
  Description <- NULL #necessary to remove NOTE
  
  ## now split NREMPs that are too long
  for (zz in 1:length(toolong)){
    cycs <- which(data$CycleStart == "NREMP" | data$CycleStart == "REMP")
    curr_toolong <- toolong[c(zz)]
    
    message(paste0("Attempting to split NREMP ", as.character(zz), " out of ", as.character(length(toolong)), "."))
    
    beg_end <- c(cycs[which(cycs==curr_toolong)], cycs[which(cycs==curr_toolong)+1]) #find beginning and end of NREMP that is too long
    
    # find RWN12 episodes that are > 12min within detected NREMP -> "lightening of sleep" = potential splitting points
    RWN12s <- which(data$Descr2 == "RWN12") #which 30s epochs are R, W, or N1/2
    RWN12s <- RWN12s[c(RWN12s>=beg_end[1] & RWN12s<=beg_end[2])] # search for R, W, N1, or N2 epochs between beginning and end of current NREMP
    RWN12s_start <- NA
    for (kk in 1:(length(RWN12s)-23)){
      if (all(seq(RWN12s[kk],length.out = 24) == RWN12s[seq(kk,kk+23)])){ # check if the sequence of R, W, N12 epochs is continuous min. 12min
        RWN12s_start <- c(RWN12s_start, RWN12s[kk])
      }
    }
    RWN12s_start <- RWN12s_start[-c(1)] #first was NA, remove
    rm(kk)
    
    # do not consider 12min period as potential splitting point if it marks the beginning of the NREMP to be split
    RWN12s_start2 <- RWN12s_start[c(RWN12s_start > beg_end[1])]
    
    if (length(RWN12s_start2) > 0){ #only split if there is a lightening of sleep following the onset of the NREMP
      # find beginnings of >12min RWN12 sequences of 'lighter' sleep
      RWN12s_start2 <- RWN12s_start[1]
      for (kk in 1:(length(RWN12s_start)-1)){
        if ((RWN12s_start[kk+1]-RWN12s_start[kk])>1){
          RWN12s_start2 <- c(RWN12s_start2, RWN12s_start[kk+1]) #if there is a discontinuity in the sequence, mark the beginning of a new 12min RWN12 sequence
        }
      }
      
      # find N3 episodes within period to split
      N3s <- which(data$Descr2 == "N3") #which are N3 epochs?
      N3s <- N3s[c(beg_end[1] < N3s) & c(beg_end[2] > N3s)]
      N3s <- N3s[c(N3s > RWN12s_start2[1])] # only N3s after start of 12min of R/W/N1/N2
      
      if (length(N3s) > 0){
        if(length(N3s) == 1){
          N3_start <- N3s
        }else{
          # find starting points of continuous N3 sequences
          N3_start <- N3s[1]
          for (kk in 1:(length(N3s)-1)){
            if ((N3s[kk+1]-N3s[kk])>1){
              N3_start <- c(N3_start, N3s[kk+1]) #if there is a discontinuity in the sequence, mark the beginning of a new NREM sequence
            }
          }
        }
        
        # select starting points of N3 following 12min of R/W/N1/2
        RWN12s_start2 <- RWN12s_start2[RWN12s_start2<tail(N3_start,1)] # RWN12s episode must start before start of last N3
        n <- NA
        for (zzz in 1:length(RWN12s_start2)){
          minpositive = function(x) min(x[x > 0])
          val <- which(minpositive(N3_start - RWN12s_start2[zzz]) == (N3_start - RWN12s_start2[zzz]))
          n <- c(n, val)
        }
        n <- na.omit(n)
        
        # added 16/11/20: do not consider first N3 within NREMP
        if (N3_start[1] == N3s[1]){ #N3_start contains starting points of continuous N3 sequences within NREMP to split
          n <- n[-c(1)]
        }
        N3_start2 <- N3_start[n] 
        
        # second NREMP starts with N3 following 12min of R/W/N1/2
        if(length(N3_start2)>0){
          data$CycleStart[c(N3_start2)] <- "NREMP"
          
          splits <- N3_start2 # all potential splitting points
          splits <- unique(splits)
          
          # ask user if s/he is happy with the result of the splitting
          # plot results
          dfplot <- data
          dfplot$time <- seq(1, nrow(dfplot)) # gives epochs
          dfplot$Description[dfplot$Description == 1] <- -1
          dfplot$Description[dfplot$Description == 2] <- -2
          dfplot$Description[dfplot$Description == 3] <- -3
          dfplot$Description[dfplot$Description == 5] <- 1
          
          pp <- ggplot(dfplot, aes(x=time, y=Description, colour=Description)) 
          pp <- pp + theme_bw()+
            geom_point() +
            geom_line(aes(x=time, y=Description))+
            ggtitle(as.character(filename))+
            xlab("Epoch") +
            ylab("Sleep Stage")+
            scale_y_continuous(limits = c(-3,2), breaks = c(-3, -2, -1, 0, 1), labels = c("N3", "N2", "N1", "W", "REM"))+
            scale_color_viridis(name = "Sleep Stage", option = "D")+
            geom_vline(xintercept = c(splits), lty = 2, colour = "red")+
            theme(legend.position = "none")+
            annotate(geom="text", x = 500, y = 2, label = paste("can split at epoch(s):", paste(splits, collapse = ","), sep = " "))+
            theme(plot.margin = unit(c(.5,.5,.5,.5), "cm"))
          print(pp)
          
          # check if both NREM parts would still be >=20min 
          part1 <- splits-beg_end[1]
          part2 <- beg_end[2]-splits+1
          print(paste0("When splitting at 1st/ 2nd/... suggestion, NREMP1 would be ", (as.character(part1/2)), " min.", 
                       "and NREMP2 would be ", (as.character(part2/2)), " min.", sep = " "))
          if (any(part1 < 30) | any(part2 < 30)){
            message("BE CAREFUL: Splitting might result in NREM period that is shorter than 15 min.")
          }
          
          # Ask about happiness level
          val <- readline("Where do you want to split? Type the 1/2/3/... to select a suggestion, n to split at a specific epoch or not at all, or skip to skip this night for now. ") #  1 splits at 1st suggested epoch, n offers to choose, skip skips this night
          
          if(val == "skip"){
            message("This period/night is skipped.")
            next
          }else if (val == "n"){
            newperiod <- readline("At which epoch do you want to start the new NREM period instead? Please type epoch number or NA to not split. ") 
            data$CycleStart[c(N3_start)] <- NA #rm old indicator
            if (newperiod != "NA"){
              data$CycleStart[as.numeric(newperiod)] <- "NREMP"
            }
          }else if (val == "1" | val == "2" | val == "3" | val == "4" | val == "5" | val == "6" | val == "7" | val == "8" | val == "9" | val == "10"){
            data$CycleStart[c(N3_start2)] <- NA #rm old indicator
            y <- as.numeric(val)
            data$CycleStart[N3_start2[y]] <- "NREMP"
          }else{
            message("Missing entry. This night is skipped.")
          }
          rm(dfplot, pp)
        }else{
          message("Cannot split. No N3 following 'lightening' of sleep has been detected.")
        }
      }else{
        message("Cannot split. No N3 following 'lightening' of sleep has been detected.")
      }
    }else{
      message("No 'lightening' of sleep or N3 detected after the onset of the NREMP. Cannot split.")
    }
  }
  return(data)
}
)


R/SleepCycles.R
#' @title Sleep Cycle Detection  
#'
#' @description Sleep cycles are largely detected according to the originally proposed criteria by Feinberg & Floyd (1979) and as described in Blume & Cajochen (2020) \doi{10.31219/osf.io/r2q8v} from sleep staging results. 
#' NREM periods are periods starting with N1 (default) or N2 at the beginning of the night and W or another NREM stage following a REM period. NREMPs have a minimal duration of 15min (can include W, up to <5min REM, except for the first REMP, 
#' for which there is no minimum duration criterion). REM following a NREM period always represents a potential REM period (REMP), however any REMP must be at least
#' 5min (except the first REMP, for which no minimum duration criterion is applied). If a NREMP exceeds 120min in duration (excl. wake), it can be split into 2 parts. 
#' The new cycle then starts with the first N3 episode following a phase (>12min) with any other stage than N3, that is
#' a lightening of sleep (cf. Rudzik et al., 2020; Jenni et al., 2004; Kurth et al., 2010). The code makes suggestions where
#' splitting could take place according to the criteria and visualises the potential splitting points on top of a hypnogram. The user can then interactively choose where to split the NREMP. However, the code also offers the possibility to provide a numeric value for an epoch 
#' at which to split or you can decide to not split at all. A combination of a NREMP and the following REMP represents one sleep cycle, except for the case when a NREMP is split.
#' In this case, the first of the two resulting NREMPs represents a sleep cycle (without REM).
#' 
#' The function requires any sleep staging results file with a column, in which the sleep stages are coded 
#' in the usual 0,1,2,3,5 (i.e., W, N1, N2, N3, REM) pattern (i.e., a numeric vector). The user can define other integers to be handled as W or N3
#' (i.e. in the case stagings were done according to the Rechtschaffen and Kales criteria including S3 and S4). The presence of further columns in the data is not an issue.
#' Staging must be in 30s epochs. Besides text files, it can also handle csv files and marker files for the Brain Vision Analyzer (filetype = "txt" (default), "csv", or "vmrk").

#' @details Besides sleep cycles (NREM-REM), the result also splits the NREM and REM parts of each cycle in percentiles. 
#' In case the length of a period is not divisible by 10 (e.g., 203 epochs), one epoch is added to percentiles in a randomized fashion to reach the correct 
#' length of a period (here: 7 percentiles of 20 epochs, 3 of 21 epochs).
#' 
#' The code offers to choose whether incomplete periods should be removed at the end of the night (rm_incomplete_period, default = F). Incomplete periods are defined by cycles that are followed
#' by <5min NREM or W (e.g. because a participant is woken up).
#'
#' Although this is not encouraged, for some participants it may be necessary to decrease the minimum duration of REM from 5min to 4 or 4.5min
#' as otherwise a seemingly 'clear' REM period is skipped. While the default length of REMPs is 10 segments, it can be decreased.
#'
#' The user can either process all files in a given directory (default) or specific files by specifying a vector of files.
#'
#' By default, the function produces and saves a plot for visual inspection of the results.
#' 
#' @references Feinberg, I. and Floyd, T.C. (1979), Systematic Trends Across the Night in Human Sleep Cycles. Psychophysiology, 16: 283-291. https://doi.org/10.1111/j.1469-8986.1979.tb02991.x
#' @references Rudzik, F., Thiesse, L., Pieren, R., Heritier, H., Eze I.C., Foraster, M., Vienneau, D., Brink, M., Wunderli, J.M., Probst-Hensch, N., Roeoesli, M., Fulda, S., Cajochen, C. (2020). Ultradian modulation of cortical arousals during sleep: effects of age and exposure to nighttime transportation noise. Sleep, Volume 43, Issue 7. https://doi.org/10.1093/sleep/zsz324
#' @references Jenni, O.E., Carskadon, M.A. (2004). Spectral Analysis of the Sleep Electroencephalogram During Adolescence. Sleep, Volume 27, Issue 4, Pages 774-783. https://doi.org/10.1093/sleep/27.4.774
#' @references Kurth, S., Ringli, M., Geiger, A., LeBourgeois, M., Jenni, O.G., Huber, R. (2010). Mapping of Cortical Activity in the First Two Decades of Life: A High-Density Sleep Electroencephalogram Study. Journal of Neuroscience. 30 (40) 13211-13219; DOI: 10.1523/JNEUROSCI.2532-10.2010 
#'
#' @param p character vector indicating the directory containing the sleep staging files
#' @param sleepstart character vector indicating whether the first NREMP at the beginning of the night should start with N1 or N2. Default: N1
#' @param files numeric vector indicating which files in 'p' to process. Default: NA
#' @param filetype character indicating file type of the files containing the sleep staging results. Can be "txt" (default) or "csv", or "vmrk" (i.e., marker files for Brain Vision Analyzer Software).
#' @param treat_as_W numeric vector indicating which values should be treated as 'wake'. Default: NA
#' @param treat_as_N3 numeric vector indicating which values should be treated as 'N3'. Default: NA
#' @param rm_incomplete_period logical: should incomplete period at the end of the night be removed? Default: FALSE.
#' @param plot logical: should a plot for the result of the detection procedure be generated and saved? Default: TTRUE.
#' @param REMP_length numeric value specifying the minimum duration of a REM period following the first REM period. Default is 10 segments (i.e. 5 minutes). Decreasing the min. length is not encouraged and should only be done following careful consideration.
#'
#' @return Saves results of the detection in a results folder in 'p'. The resulting textfile contains the sleepstages in a column named 'SleepStages', the sleep cycles in 
#' a column 'SleepCycles' (numeric value indicating the cycle number), information on whether it is a NREM or REM period (numeric value in column 'N_REM', 0 = NREM, 1 = REM), and an indicator of the percentiles
#' of the (N)REM period of the cycle (numeric value in 'percentile' column; 1 = first percentile, 2 = second percentile, etc.). In case a (N)REM period is less than 10 epochs long,
#' no percentiles are calculated (all epochs are coded as '1' in the 'percentile' column).
#'
#' @import ggplot2 reshape2 plyr stringr viridis
#' 
#' @importFrom stats na.omit time
#' @importFrom utils data glob2rx head read.csv read.table tail write.table
#' 
#' @examples 
#' data(sleepstages)
#' olddir <- getwd()
#' newdir <- file.path(tempdir(),"SleepCycles_exmpl")
#' dir.create(newdir, showWarnings = FALSE)
#' write.table(sleepstages, file = paste(newdir, "sleepstages.vmrk", sep = "/"), 
#' row.names=FALSE, col.names = FALSE, quote = FALSE, sep = ",")
#' SleepCycles(newdir, filetype = "vmrk")
#' setwd(olddir)
#' 
#' \dontrun{
#' # Dataset that requires splitting of a NREMP
#' data(sleepstages2)
#' olddir <- getwd()
#' newdir <- file.path(tempdir(),"SleepCycles_exmpl2")
#' dir.create(newdir, showWarnings = FALSE)
#' write.table(sleepstages2, file = paste(newdir, "sleepstages2.txt", sep = "/"),
#'             row.names=FALSE, col.names = TRUE, quote = FALSE, sep = ",")
#' SleepCycles(newdir, filetype = "txt")
#' setwd(olddir)
#' }
#'
#' @export
SleepCycles <- function(p, sleepstart = "N1", files = NA, filetype = "txt", treat_as_W = NA, treat_as_N3 = NA, rm_incomplete_period = FALSE, plot = TRUE, REMP_length = 10){

  # # --- set a few things
  oldwd <- getwd()
  on.exit(setwd(oldwd))
  
  setwd(p)
  filename <- NA
  REMs <- NA
  Description <- NA

  # check if there are result files of this function in the directory as they will mess with the code
  # stop code execution if they are found
  x <- list.files(p, pattern = glob2rx("*SCycles.txt"))
  if (length(x)>0){
    stop("Please remove files from previous Sleep Cycle detections from the folder.")
  }
  
  #----- list all files in directory
  if (filetype == "vmrk"){
    d <- list.files(p, pattern = "*.vmrk")
    hd <- NA
  }else if (filetype == "txt"){
    d <- list.files(p, pattern = "*.txt")
    hd <- readline("Do your files have a header with column names (y/n)? ") #check if first line contains column names
    sp <- readline("Which separator do the files have? Choose one of the following: NA or , or ; or tabulator. If the data only contains one column, write NA.") #check which separator is used
    if (sp == "tabulator"){
      sp = "\t"
    }
    if (sp == "NA"){
      sp = ""
    }
  }else if (filetype == "csv"){
    d <- list.files(p, pattern = "*.csv") # csv files were added on 17/02/21
    hd <- readline("Do your files have a header with column names (y/n)? ") #check if first line contains column names
    sp <- readline("Which separator do the files have? Choose one of the following: , or ; or tabulator. If the data only contains one column, write NA.") #check which separator is used
    if (sp == "tabulator"){
      sp = "\t"
    }
    if (sp == "NA"){
      sp = ""
    }
  }
  
  #----- has a vector for a subset of files to be processed been specified?
  if (!all(is.na(files))){
    d <- d[files]
  }
  
  #----- prepare results folder, create new directory
  sv <- paste("SleepCycles", Sys.Date(), sep = "_")
  dir.create(file.path(paste(p, sv, sep = "/")), showWarnings = FALSE)
  
  #--------------------------------------------------------
  #----- loop through files to determine sleep cycles
  #--------------------------------------------------------
  
  for (i in 1:length(d)){
    print(i) #tell user which file is processed
    filename <- d[i]
    
    ## load data
    D <- load_data(filetype, filename, treat_as_W, treat_as_N3, hd, sp)
    data <- D[[1]]
    cycles <- D[[2]]
    rm(D)
    
    ##-- prep data for further processing
    data <- prep_data(data, treat_as_W, treat_as_N3)
    
    # Find NREM periods: start with N1 and can then also include W. >=15min
    NREMWs <- which(data$Descr3 == "NREM"| data$Descr3 == "W") #which 30s epochs are NREM or wake
    NREMs <- which(data$Descr3 == "NREM")
    first_N2 <- which(data$Description == 2)[1] # for option to have first NREMP start with N2
    
    # exclude W or W/N1 at the beginning of the night before the first NREM epoch
    if (sleepstart == "N1"){
      NREMWs <- subset(NREMWs, NREMWs >= NREMs[1]) 
    }else if (sleepstart == "N2"){
      NREMWs <- subset(NREMWs, NREMWs >= first_N2) 
    }
    
    ## Loop through NREMWs 
    # check if the sequence of NREWM is continuous and the period is >=15min AND beginning is not wake -> first NREMP
    # Further: find discontinuities in the sequence (= potential beginnings of new NREM periods during the remaining night)
    NREMWs_start2 <- find_NREMPs(NREMWs, data)
    data$CycleStart <- NA
    data$CycleStart[NREMWs_start2] <- "NREMP" #marks all potential NREM period beginnings
    
    ## Find REM episodes (first can be <5min, others have to be at least 5min)
    REMs_start2 <- find_REMPs(REMs, REMP_length, data)
    data$CycleStart[REMs_start2] <- "REMP"
    
    ## remove several NREMPs or REMPs in a row
    rm <- delete_reps(data)
    data$CycleStart[c(rm)] <- NA
    
    ## is any NREM part (excl. wake) of a NREMP longer than 120min?
    toolong <- is.toolong(data)
    toolong1 <- toolong # for comparison with second round
    ## now split NREMPs that are too long
    if (length(toolong) > 0){
      data <- toolong_split(data, toolong, filename)
    }
    
    ## now check again if there are still NREMPs > 120min
    ## is any NREM part (excl. wake) of a NREMP longer than 120min?
    rm(toolong)
    toolong <- is.toolong(data)
    toolong2 <- toolong
    
    ## now split NREMPs that are too long
    if (length(toolong) > 0 & any(toolong2 != toolong1)){
      message("~ Still detected a NREMP > 120min. Let's go through the splitting process again. ~")
      data <- toolong_split(data, toolong, filename)
    }

    # -----------------------------------------------------------------------------
    # now finish and add cycle markers and percentiles
    
    # add cycle markers to file (only for NREM cycles) & add NREM vs. REM part info
    data <- addinfo1(data)
    
    # remove incomplete NREM-REM cycle at the end of the night (i.e., cycles followed by <5min NREM or W)
    if (rm_incomplete_period == TRUE){
      data  <- rm.incompleteperiod(data)
    
    #remove NREM/W following last REMP (in case no new NREMP begins) or REM/W following last NREMP  (in case no new REMP begins)
    }else if (rm_incomplete_period == FALSE){
      data <- clean_endofnight(data)
    }
    
    ## merge cycle marker & NREM/REM marker & add percentiles of NREM & REM parts
    data <- addinfo2(data)
    
    ## prep new marker file with cycle info
    if (filetype == "vmrk"){
      cycles <- cycles[,-c(4,5)]
    }
    cycles$Description <- data$cycle_perc
    cycles$SleepCycle <- data$cycles
    cycles$N_REM <- data$REM.NREM
    cycles$percentile <- data$perc
    
    ## save new file
    svv <- paste(p, sv, sep = "/")
    name <- unlist(stringr::str_split(filename, pattern = "[.]"))[1]
    savename <- paste(name, "SCycles.txt", sep = "_")
    write.table(cycles, file = paste(svv, savename, sep = "/"), row.names = F)
    
    ## plot results if desired
    if (plot == TRUE){
      plot_result(data, filetype, name, svv)
    }
  }
}
